### ISSUE TITLE
Sign with SIGHASH_ANYONECANPAY

### ISSUE THREAD

**Issue by petertodd** [JoinMarket-Org/joinmarket#599]
This would make it easy to add new inputs to joinmarket transactions to bump fees. Importantly, anyone would be able to add those inputs - all participants in the join, and indeed, anyone at all. The usual argument against doing this is transaction mutability, but it's not safe in Joinmarket to spend unconfirmed change for both privacy and security reasons anyway.

This is most effective if joinmarket transactions set the opt-in RBF flag, although it's not needed; in cases where a transaction is taking too long to mine it's not hard at all to broadcast new versions of transactions as in that circumstance chances are lots of nodes don't have the tx in their mempool anyway. Meanwhile, broadcasting a new tx with an additional input leaks less information. Of course, you have to happen to have an input of appropriate (small) size, but if you do have one, there's no reason not to use it; you can always use a second transaction to create such an input too. (kinda like CPFP, but much more reliable as CPFP often fails due to the low-fee tx not propagating)

SIGHASH_ANYONECANPAY will act to distinguish joinmarket txs from others... but let's face it, joinmarket txs are already easily identifiable, so I don't think that's a major concern.

If a reasonable number of ANYONECANPAY transactions existed on the network there's another advantage: since anyone can bump fees, this provides some plausible deniability for the ownership of low-value inputs. Anyone can get rid of dust outputs by donating them to others, effectively coinjoining their funds with others in that transaction. This also makes it harder for 51% attackers to do transaction rollback attacks, like were recently suggested as a response to the Bitfinex attack.

In short, this won't get used _too_ much, but the downside of doing this is minimal for joinmarket, and it has some important upsides.

**Comment by chris-belcher** [comment]
Interesting thought, thanks for that.

> This also makes it harder for 51% attackers to do transaction rollback attacks, like were recently suggested as a response to the Bitfinex attack.

Could you explain this a bit more? How does ANYONECANPAY do that.

**Comment by petertodd** [comment]
Basically, what you want is for all transactions to be linked to all other transactions, so attempting to selectively reverse some transactions isn't possible because others depend on them; you want censorship to be an all-or-nothing proposition.

For example, if I have a 2BTC output, and send 1.998BTC to someone else with a 0.001BTC fee, that leaves me with a 0.001BTC output. If I take someone elses transaction and add my 0.001BTC output to it, I've both donated my funds to that second transaction - possibly making it confirm a little faster if it was low fee - but I've also made it depend on mine. Now when the output of that second transaction is spent by a third transaction, it'll depend on both transactions - change either one and the third transaction is invalid. So any miner trying to censor the first transaction in a reorg attack is forced to censor all three - a rather disruptive thing to do, with a lot of collateral damage!

ANYONECANPAY makes all this possible, basically because signatures using it are only signing the outputs they want in the transaction, and _one_ input - other inputs can be added at will without making the signature invalid.

**Comment by adlai** [comment]
At first glance, this seems to require knowing in advance the size and number of the resulting transaction's inputs and outputs. Guessing too low reduces the amount of fee savings, whereas guessing too high could leave the taker waiting indefinitely for extra participants (or filling the rest of the transaction alone).

**Comment by AdamISZ** [comment]
I agree that the watermarking effect doesn't matter today, and arguably won't in future; I see two possible classes of false positives for JM txes; 1 is already existing - some that look like 3 party JM txes as identified by @adlai with cjhunt, but fairly obviously weren't, and in future (now?) spoofed JM txes a la what was proposed by Samourai. Aiming for some heterogeneous standard tx format into which JM could fit seems like a really tall order.

It's also worth mentioning in case anyone reading is unaware, that we have dynamic fees in JM and rarely if ever get "stuck" transactions any more (at least I haven't seen any reports for a long time). So txes not fully propagated is not a scenario we realistically have right now.

> At first glance, this seems to require knowing in advance the size and number of the resulting transaction's inputs and outputs. Guessing too low reduces the amount of fee savings, whereas guessing too high could leave the taker waiting indefinitely for extra participants (or filling the rest of the transaction alone).

I take your point but I guess it depends on if we take up @petertodd 's opt-in RBF suggestion; if we did, JM txs could be done exactly as today, i.e. assuming no attached dust outputs, but just allowing them by flagging ANYONECANPAY on the sigs. I have no practical knowledge of opt-in RBF so I'm just assuming it as a black box that automagically succeeds in replacing unconf transactions with fee bumps under any situation (I guess that's not really true?).

At a more nitty-gritty practical level I don't know how easy it would be to deal with replacement of transactions, our crude walletnotify method seems to work OK as is, but this would bump up the sophistication a notch.

**Comment by petertodd** [comment]
@adlai Why do you think allowing new inputs to be added after-the-fact is any worse than the status quo? Right now if you guess the wrong fee your tx is stuck, and there's little you can do about it.

@AdamISZ So, re: replacement, remember that we already have to deal with that case - and do deal with that case - because counterparties sometimes double-spend transactions. I've seen this happen personally multiple times, and nothing bad happens other than of course the transaction getting cancelled (I'm using JM with Bitcoin Core FWIW).

As for opt-in-RBF, for fee bumps larger than the network-spam-threshold, yes, it does succeed in doing replacements most of the time - 100% of the time if a transaction is stuck due to low-fee. There is a disadvantage that some wallets - notably Coinbase - misleadingly treat opt-in-RBF as more risky than transactions not using it however, so we may want to leave that off for the first version where this change is made.

With regard to fingerprinting, I'll also be submitting a patch to Bitcoin Core to implement this change when we have spend-unconfirmed-change disabled, so JM won't be the only wallet doing this. Additionally, GreenAddress comes to mind as another wallet that can do this, as it too doesn't spend unconfirmed change.

Dynamic fees aren't a panacea: fundamentally you can't predict the future, so while infrequent, there are times when fee estimation turns out to be wrong.

**Comment by AdamISZ** [comment]
@petertodd re: re: replacement, yes of course you're right, I forgot about that.

**Comment by petertodd** [comment]
Opened a pull-req to implement this on Bitcoin Core: https://github.com/bitcoin/bitcoin/pull/8543
